filelog.go

```diff
@@ -345,6 +357,117 @@ func (w *FileLogWriter) archiveFiles(dir string) error {
+		return
+	}
+
+	// Defer closing the underlying file
+	defer func() {
+		err = compressedFile.Close()
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		}
+	}()
+
+	var compressedFileWriter io.Writer
+	switch method {
+	case "gz":
+		compressedFileWriter = gzip.NewWriter(compressedFile)
+	case "zip":
+		// zipWriter is the outer container, compressedFileWriter is an entry inside the zip
+		zipWriter := zip.NewWriter(compressedFile)
+
+		// Defer closing the zip writer
+		defer func() {
+			err = zipWriter.Close()
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't close zip writer on %q: %s\n", w.filename, compressedInprogressFilename, err)
+			}
+		}()
+
+		compressedFileWriter, err = zipWriter.Create(basename)
+		if err != nil {
+			fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't open zip file entry: %s\n", w.filename, err)
+			return
+		}
+	default:
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Unknown compression method: %q\n", w.filename, method)
+		return
+	}
+
+	// Defer closing the writer, but only if we have a io.WriteCloser
+	defer func() {
+		// gzip compression requires closure, but zip doesn't have a Close() method
+		writeCloser, ok := compressedFileWriter.(io.WriteCloser)
+		if ok {
+			err = writeCloser.Close()
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't finalize compression on %q: %s\n", w.filename, compressedInprogressFilename, err)
+			}
+		}
+	}()
+
+	// If we return early after this point, we need to cleanup the .inprogress file
+	cleanupInprogress := true
+	defer func() {
+		if cleanupInprogress {
+			err = os.Remove(compressedInprogressFilename)
+			if err != nil {
+				fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't remove temporary file %q: %s\n", w.filename, err)
+			}
+		}
+	}()
+
+	// Read plain file, write to compressed file
+	_, err = io.Copy(compressedFileWriter, plainFile)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't write compressed file %q: %s\n", w.filename, compressedInprogressFilename, err)
+		return
+	}
+
+	// Since all we do is read from plainFile, we can ignore any error closing it
+	_ = plainFile.Close()
+
+	// Rename compressed file
+	err = os.Rename(compressedInprogressFilename, compressedFilename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't rename file %q to %q: %s\n", w.filename, compressedInprogressFilename, compressedFilename, err)
+		return
+	}
+
+	// Upon successfully renaming the file, we no longer need to cleanup the .inprogress file
+	cleanupInprogress = false
+
+	// Delete plain file
+	err = os.Remove(filename)
+	if err != nil {
+		fmt.Fprintf(w.errorWriter, "FileLogWriter(%q): Couldn't remove file %q: %s\n", w.filename, filename, err)
+
+		// If we can't remove the old plain file, delete the compressed file so we don't affect the
```

Comment:
I think we have different views about how compression and retention should function. In your scenario, how would we handle retention if the program was offline for two weeks and we were configured to keep 30 days? Should we delete two weeks of logs? If we went offline for more than 30 days (or the retention period), wouldn't we clear the old logs on first start and potentially prevent ourselves from learning the reason the program stopped? A similar situation could occur if the system time drifts substantially and we jump forward suddenly.

In my view, I'd rather keep a rolling buffer of the most recent 30 files rather than exactly 30 days. Likewise, if compressions fails, I'm tempted to leave it alone forever, and never reattempt. Compression can burn a lot of CPU in the case of large log files, and if it failed the first time (e.g. due to a nearly full disk), what stops us from going into a loop where we attempt compression multiple times, only to fail due to the disk? 

I agree that cleaning up inprogress files at startup would be an improvement.